backend/oasst_backend/journal_writer.py

```diff
@@ -0,0 +1,122 @@
+    text_reply_to_post = "text_reply_to_post"
+    post_rating = "post_rating"
+    post_ranking = "post_ranking"
+
+
+@payload_type
+class JournalEvent(BaseModel):
+    type: str
+    person_id: Optional[UUID]
+    post_id: Optional[UUID]
+    workpackage_id: Optional[UUID]
+    task_type: Optional[str]
+
+
+@payload_type
+class TextReplyEvent(JournalEvent):
+    type: Literal[JournalEventType.text_reply_to_post] = JournalEventType.text_reply_to_post
+    length: int
+    role: str
+
+
+@payload_type
+class RatingEvent(JournalEvent):
+    type: Literal[JournalEventType.post_rating] = JournalEventType.post_rating
+    rating: int
+
+
+@payload_type
+class RankingEvent(JournalEvent):
+    type: Literal[JournalEventType.post_ranking] = JournalEventType.post_ranking
+    ranking: list[int]
+
+
+class JournalWriter:
+    def __init__(self, db: Session, api_client: ApiClient, person: Person):
+        self.db = db
+        self.api_client = api_client
+        self.person = person
+        self.person_id = self.person.id if self.person else None
+
+    def log_text_reply(self, work_package: WorkPackage, post_id: UUID, role: str, length: int) -> Journal:
+        return self.log(
+            task_type=work_package.payload_type,
+            event_type=JournalEventType.text_reply_to_post,
+            payload=TextReplyEvent(role=role, length=length),
+            workpackage_id=work_package.id,
+            post_id=post_id,
+        )
+
+    def log_rating(self, work_package: WorkPackage, post_id: UUID, rating: int) -> Journal:
+        return self.log(
+            task_type=work_package.payload_type,
+            event_type=JournalEventType.post_rating,
+            payload=RatingEvent(rating=rating),
+            workpackage_id=work_package.id,
+            post_id=post_id,
+        )
+
+    def log_ranking(self, work_package: WorkPackage, post_id: UUID, ranking: list[int]) -> Journal:
+        return self.log(
+            task_type=work_package.payload_type,
+            event_type=JournalEventType.post_ranking,
+            payload=RankingEvent(ranking=ranking),
+            workpackage_id=work_package.id,
+            post_id=post_id,
+        )
+
+    def log(
+        self,
+        *,
+        payload: JournalEvent,
+        task_type: str,
+        event_type: str = None,
+        workpackage_id: Optional[UUID] = None,
+        post_id: Optional[UUID] = None,
+        commit: bool = True,
+    ) -> Journal:
+        if event_type is None:
+            if payload is None:
+                event_type = "null"
+            else:
+                event_type = type(payload).__name__
+
+        if payload.person_id is None:
+            payload.person_id = self.person_id
+        if payload.post_id is None:
+            payload.post_id = post_id
+        if payload.workpackage_id is None:
+            payload.workpackage_id = workpackage_id
+        if payload.task_type is None:
+            payload.task_type = task_type
```

Comment:
You can write it in a shorter way : 
```python 
if event_type is None:
    event_type = "null" if payload is None else type(payload).__name__

payload.person_id = payload.person_id or self.person_id
payload.post_id = payload.post_id or post_id
payload.workpackage_id = payload.workpackage_id or workpackage_id
payload.task_type = payload.task_type or task_type
```

Reply:
Thank for the tip. Just to keep it short I could even also write 
`event_type = payload and type(payload).__name__ or "null"` 
It is a matter of taste ... I am personally not the biggest fan of the python ternary operator.